hub/core/query/autocomplete.py

```diff
@@ -0,0 +1,289 @@
+            ht["type"] = "OP"
+        hubtokens.append(ht)
+    return hubtokens
+
+
+def _parse_no_fail(s: str, ds: hub.dataset):
+    try:
+        return _parse(s, ds)
+    except TokenError:
+        return _parse(s[:-1], ds)
+
+
+def _sort_suggestions(s: List[dict]) -> None:
+    s.sort(key=lambda s: s["string"])
+
+
+def _initial_suggestions(ds):
+    tensors = [{"string": k, "type": "TENSOR"} for k in ds._ungrouped_tensors]
+    _sort_suggestions(tensors)
+    groups = [{"string": k, "type": "GROUP"} for k in ds._groups_filtered]
+    _sort_suggestions(groups)
+    groups.sort(key=lambda s: s["string"])
+    return tensors + groups
+
+
+def _tensor_suggestions(ds, tensor):
+    methods = [{"string": m, "type": "METHOD"} for m in _TENSOR_METHODS]
+    _sort_suggestions(methods)
+    properties = [{"string": p, "type": "PROPERTY"} for p in _TENSOR_PROPERTIES]
+    _sort_suggestions(properties)
+    return methods + properties
+
+
+def _const_suggestions():
+    return [{"string": str(v), "type": "CONSTANT"} for v in (True, False, None)]
+
+
+def _group_suggestions(ds, group):
+    return [{"string": k, "type": "TENSOR"} for k in group._ungrouped_tensors] + [
+        {"string": k, "type": "GROUP"} for k in group._groups_filtered
+    ]
+
+
+def _parse_last_tensor(tokens, ds):
+    tensor = []
+    n = len(tokens) - 1
+    if tokens[n]["type"] in ("UNKNOWN", "OP"):
+        n -= 1
+    for i in range(n, -1, -1):
+        if tokens[i]["string"] == ".":
+            continue
+        elif tokens[i]["type"] in ("TENSOR", "GROUP"):
+            tensor.append(tokens[i]["string"])
+        else:
+            break
+    ret = ds
+    for i in range(len(tensor) - 1, -1, -1):
+        ret = ret[tensor[i]]
+    return ret
+
+
+def _filter(suggestions: List[dict], string: str):
+    return list(
+        filter(lambda x: x["string"].lower().startswith(string.lower()), suggestions)
+    )
+
+
+def _autocomplete_response(
+    suggestions: List[dict], tokens: List[dict], replace: str = ""
+):
+    return {
+        "suggestions": suggestions,
+        "tokens": tokens,
+        "replace": replace,
+    }
+
+
+def _prefix_suggestions(suggestions: List[dict], prefix: str) -> List[dict]:
+    return list(
+        map(lambda s: {"string": prefix + s["string"], "type": s["type"]}, suggestions)
+    )
+
+
+def _op_suggestions():
+    ops = [
+        "==",
+        ">",
+        "<",
+        ">=",
+        "<=",
+        "!=",
+    ]
+    return [{"string": op, "type": "OP"} for op in ops]
+
+
+def autocomplete(s: str, ds: hub.dataset) -> dict:
+    if not s.strip():
+        return _autocomplete_response(_initial_suggestions(ds), [])
+    try:
+        tokens = _parse(s, ds)
+    except TokenError:
+        return _autocomplete_response([], _parse_no_fail(s[:-1], ds))
+
+    last_token = tokens[-1]
+    last_type = last_token["type"]
+    last_string = last_token["string"]
+
+    if last_type == "UNKNOWN":
```

Comment:
Can we have better structure for this ? Perhaps, ```Dict[Token, Func]``` ?
Or at least documentation of what is going on here ?